File: model/video-section.go
package model

import (
	"fmt"
	"gorm.io/gorm"
)

type VideoSection struct {
	gorm.Model

	Description  string `gorm:"not null;index:,class:FULLTEXT" json:"description"`
	StartHours   uint   `gorm:"not null" json:"startHours"`
	StartMinutes uint   `gorm:"not null" json:"startMinutes"`
	StartSeconds uint   `gorm:"not null" json:"startSeconds"`

	StreamID uint `gorm:"not null" json:"streamID"`
	FileID   uint `gorm:"not null" json:"fileID"`
}

func (v VideoSection) TimestampAsString() string {
	if v.StartHours == 0 {
		return fmt.Sprintf("%02d:%02d", v.StartMinutes, v.StartSeconds)
	}
	return fmt.Sprintf("%02d:%02d:%02d", v.StartHours, v.StartMinutes, v.StartSeconds)
}

File: web/ts/VideoSection.ts
import { Delete, postData, Section } from "./global";

export class VideoSection {
    private readonly streamID: number;

    existingSections: Section[];
    newSections: Section[];
    current: Section;
    unsavedChanges: boolean;

    constructor(streamID) {
        this.newSections = [];
        this.existingSections = [];
        this.streamID = streamID;
        this.unsavedChanges = false;
        this.resetCurrent();
    }

    load() {
        return fetch(`/api/stream/${this.streamID}/sections`)
            .then((res) => res.json())
            .then((sections) => {
                if (sections === undefined || sections === null) {
                    this.existingSections = [];
                } else {
                    this.existingSections = sections;
                }
            });
    }
    pushNewSection() {
        this.current.friendlyTimestamp = VideoSection.timeStringAsString(this.current);
        this.newSections.push({ ...this.current });
        this.resetCurrent();
        this.unsavedChanges = true;
    }
    removeNewSection(section: Section) {
        this.newSections = this.newSections.filter((s) => s !== section);
        this.unsavedChanges = true;
    }
    publishNewSections() {
        postData(`/api/stream/${this.streamID}/sections`, this.newSections).then(async () => {
            await this.load(); // load sections again to avaid js-sorting
            this.newSections = [];
        });
        this.unsavedChanges = false;
    }
    removeExistingSection(id: number) {
        Delete(`/api/stream/${this.streamID}/sections/${id}`).then(async () => {
            await this.load();
        });
    }
    private static timeStringAsString(section: Section): string {
        let s = "";

        if (section.startHours > 0) {
            s += section.startHours;
            s += ":";
        }
        if (section.startMinutes < 10) {
            s += `0${section.startMinutes}`;
        } else {
            s += section.startMinutes;
        }
        s += ":";
        if (section.startSeconds < 10) {
            s += `0${section.startSeconds}`;
        } else {
            s += section.startSeconds;
        }
        return s;
    }
    private resetCurrent() {
        this.current = {
            description: "",
            startHours: 0,
            startMinutes: 0,
            startSeconds: 0,
            streamID: this.streamID,
        };
    }
}
